File: services/indicator_calculators/orb_calculator.py
import logging
from collections.abc import Sequence
from datetime import date, datetime, timedelta
from typing import Any

import pytz

from core.types_registry import AssetClass, AssetSymbol, OHLCVBar
from services.time_utils import create_market_open_time, utc_timestamp_ms_to_et_datetime

logger = logging.getLogger(__name__)


def calculate_orb(
    bars: Sequence[OHLCVBar],
    symbol: AssetSymbol,
    or_minutes: int = 5,
    avg_period: int = 14,
    now_func: Any = None,
) -> dict[str, Any]:
    """
    Calculate ORB (Opening Range Breakout) indicator including relative volume and direction.

    This implementation extracts opening range data from 5-minute bars and calculates:
    1. Relative Volume: Current opening range volume vs average of previous periods
    2. Direction: Bullish, bearish, or doji based on opening range price movement

    Args:
        bars: List of 5-minute OHLCV bars (aggregated bars from Polygon)
        symbol: Asset symbol (used to determine opening range time based on asset class)
        or_minutes: Opening range period in minutes (default: 5)
        avg_period: Period for calculating average volume (default: 14 days)
        now_func: Optional function to use instead of datetime.now for testing

    Returns:
        Dictionary with 'rel_vol' (relative volume as percentage) and 'direction' (bullish/bearish/doji)

    Reference Paper Implementation:
        The paper uses:
        - Relative Volume = (Volume in first n minutes) / (Avg of last 14 days) * 100
        - This matches our implementation

        Note: The paper also mentions ATR filtering (> $0.50 over 14 days) but that is
        a separate filter, not part of the ORB calculation itself.
    """
    current_time_utc = datetime.now(pytz.timezone("UTC"))
    current_time_et = current_time_utc.astimezone(pytz.timezone("US/Eastern"))
    
    print("=" * 80)
    print("ORB CALCULATOR: Starting calculation")
    print("=" * 80)
    logger.info("=" * 80)
    logger.info("ORB CALCULATOR: Starting calculation")
    logger.info(f"ORB CALCULATOR: Current time (UTC): {current_time_utc}")
    logger.info(f"ORB CALCULATOR: Current time (ET): {current_time_et}")
    logger.info("=" * 80)

    if not bars or len(bars) == 0:
        logger.error("ORB CALCULATOR: No bars provided")
        return {"rel_vol": None, "direction": None, "error": "No bars provided"}

    print(f"ORB CALCULATOR: Processing {len(bars)} bars for {symbol.ticker}")
    logger.info(f"ORB Calculator: Processing {len(bars)} bars for {symbol.ticker}")
    
    # Log raw bar timestamps before conversion
    if bars:
        first_bar_raw = bars[0]["timestamp"]
        last_bar_raw = bars[-1]["timestamp"]
        logger.info(f"ORB CALCULATOR: Raw timestamps - First: {first_bar_raw}, Last: {last_bar_raw}")

    # Convert timestamps to datetime objects (assuming milliseconds)
    bar_data: list[dict[str, Any]] = []
    for bar in bars:
        # Use time_utils function to properly convert UTC timestamp to ET datetime
        dt_et = utc_timestamp_ms_to_et_datetime(bar["timestamp"])
        bar_data.append(
            {
                "timestamp": dt_et,
                "open": bar["open"],
                "high": bar["high"],
                "low": bar["low"],
                "close": bar["close"],
                "volume": bar["volume"],
            }
        )

    if bar_data:
        first_bar_time = bar_data[0]["timestamp"]
        last_bar_time = bar_data[-1]["timestamp"]
        time_diff = current_time_et - last_bar_time
        delay_minutes = time_diff.total_seconds() / 60
        
        print(
            f"ORB CALCULATOR: Converted bars. First bar: {first_bar_time}, Last bar: {last_bar_time}"
        )
        logger.info(
            f"ORB Calculator: Converted bars. First bar: {first_bar_time}, Last bar: {last_bar_time}"
        )
        logger.info(f"ORB CALCULATOR: Current time (ET): {current_time_et}")
        logger.info(f"ORB CALCULATOR: Delay from last bar: {delay_minutes:.2f} minutes")
        
        if delay_minutes < 5:
            logger.info(f"ORB CALCULATOR: ✅ Bar data appears REAL-TIME (delay < 5 minutes)")
        elif delay_minutes < 15:
            logger.warning(f"ORB CALCULATOR: ⚠️ Bar data appears SLIGHTLY DELAYED (delay {delay_minutes:.2f} minutes)")
        else:
            logger.warning(f"ORB CALCULATOR: ⚠️ Bar data appears SIGNIFICANTLY DELAYED (delay {delay_minutes:.2f} minutes)")

    # Determine today's date early for logging
    if now_func is None:
        now_func = datetime.now
    today_date = now_func(pytz.timezone("US/Eastern")).date()
    logger.info(f"ORB Calculator: Today's date: {today_date}")

    # Group bars by date
    daily_groups: dict[date, list[dict[str, Any]]] = {}
    for bar_dict in bar_data:
        bar_date = bar_dict["timestamp"].date()
        if bar_date not in daily_groups:
            daily_groups[bar_date] = []
        daily_groups[bar_date].append(bar_dict)

    print(f"ORB CALCULATOR: Grouped into {len(daily_groups)} days")
    logger.info(f"ORB Calculator: Grouped into {len(daily_groups)} days")
    logger.info(f"ORB CALCULATOR: Available dates: {sorted(daily_groups.keys())}")
    logger.info(f"ORB CALCULATOR: Today's date: {today_date}")
    
    if today_date in daily_groups:
        logger.info(f"ORB CALCULATOR: ✅ Today's data ({today_date}) is AVAILABLE")
    else:
        latest_date = max(daily_groups.keys()) if daily_groups else None
        if latest_date:
            days_behind = (today_date - latest_date).days
            logger.warning(f"ORB CALCULATOR: ⚠️ Today's data ({today_date}) is NOT AVAILABLE")
            logger.warning(f"ORB CALCULATOR: ⚠️ Latest available date: {latest_date} ({days_behind} days behind)")
        else:
            logger.warning(f"ORB CALCULATOR: ⚠️ Today's data ({today_date}) is NOT AVAILABLE - No dates found")

    # Calculate opening range volumes and directions for each day
    or_volumes: dict[date, float] = {}
    or_directions: dict[date, str] = {}

    for date_key, day_bars in daily_groups.items():
        # Sort bars by timestamp to ensure chronological order
        day_bars_sorted = sorted(day_bars, key=lambda b: b["timestamp"])

        # Determine opening range time based on asset class
        if symbol.asset_class == AssetClass.CRYPTO:
            # For crypto, use UTC midnight (00:00:00) as opening range, convert to ET
            utc_midnight = pytz.timezone("UTC").localize(
                datetime.combine(date_key, datetime.strptime("00:00", "%H:%M").time())
            )
            open_time = utc_midnight.astimezone(pytz.timezone("US/Eastern"))
        else:
            # For stocks, use 9:30 AM Eastern as opening range
            open_time = create_market_open_time(date_key, hour=9, minute=30)

        # Debug: Show first few bars for this day
        print(f"ORB CALCULATOR: Bars for {date_key} (showing first 5):")
        for i, bar in enumerate(day_bars_sorted[:5]):
            print(f"  Bar {i}: timestamp={bar['timestamp']}")

        # Look for bars within the opening range (9:30 AM to 9:35 AM)
        open_range_end = open_time + timedelta(minutes=5)

        # Find bars that start within the opening range
        or_candidates = [
            bar for bar in day_bars_sorted if open_time <= bar["timestamp"] < open_range_end
        ]

        if not or_candidates:
            print(f"ORB CALCULATOR: No opening range bar found for {date_key}")
            print(f"ORB CALCULATOR: Looking for bars between {open_time} and {open_range_end}")
            logger.warning(f"ORB Calculator: No opening range bar found for {date_key}")
            logger.warning(f"ORB CALCULATOR: Expected opening range: {open_time} to {open_range_end}")
            logger.warning(f"ORB CALCULATOR: Available bars for {date_key}: {len(day_bars_sorted)}")
            if day_bars_sorted:
                logger.warning(f"ORB CALCULATOR: First bar time: {day_bars_sorted[0]['timestamp']}")
                logger.warning(f"ORB CALCULATOR: Last bar time: {day_bars_sorted[-1]['timestamp']}")
            continue

        # Pick the earliest bar in the opening range
        or_bar = or_candidates[0]

        print(
            f"ORB CALCULATOR: Found {len(or_candidates)} opening range bars for {date_key}, using: {or_bar['timestamp']}"
        )

        # Calculate opening range metrics from single 5-minute bar
        or_volume = or_bar["volume"]
        or_open = or_bar["open"]
        or_close = or_bar["close"]

        # Determine direction
        if or_close > or_open:
            direction = "bullish"
        elif or_close < or_open:
            direction = "bearish"
        else:
            direction = "doji"

        print(
            f"ORB CALCULATOR: {date_key} - OR bar: time={or_bar['timestamp']}, volume={or_volume}, direction={direction}, open={or_open}, close={or_close}"
        )
        logger.info(
            f"ORB Calculator: {date_key} - OR bar: time={or_bar['timestamp']}, volume={or_volume}, direction={direction}, open={or_open}, close={or_close}"
        )

        or_volumes[date_key] = or_volume
        or_directions[date_key] = direction

    # Calculate relative volume
    sorted_dates = sorted(or_volumes.keys())
    if len(sorted_dates) < 1:
        return {"rel_vol": None, "direction": None, "error": "Insufficient days"}

    print(f"ORB CALCULATOR: Found {len(sorted_dates)} days with opening range bars")
    print(f"ORB CALCULATOR: Date range: {sorted_dates[0]} to {sorted_dates[-1]}")
    logger.info(f"ORB Calculator: Found {len(sorted_dates)} days with opening range bars")
    logger.info(f"ORB Calculator: Date range: {sorted_dates[0]} to {sorted_dates[-1]}")
    logger.info(f"ORB CALCULATOR: Days with OR data: {sorted_dates}")
    
    # Check if today has OR data
    if today_date in sorted_dates:
        logger.info(f"ORB CALCULATOR: ✅ Today ({today_date}) has opening range data")
        today_or_volume = or_volumes.get(today_date, 0)
        today_or_direction = or_directions.get(today_date, "unknown")
        logger.info(f"ORB CALCULATOR: Today's OR - Volume: {today_or_volume}, Direction: {today_or_direction}")
    else:
        latest_or_date = sorted_dates[-1] if sorted_dates else None
        if latest_or_date:
            days_behind = (today_date - latest_or_date).days
            logger.warning(f"ORB CALCULATOR: ⚠️ Today ({today_date}) does NOT have opening range data")
            logger.warning(f"ORB CALCULATOR: ⚠️ Latest OR date: {latest_or_date} ({days_behind} days behind)")

    # Determine target date (today or last trading day)
    if symbol.asset_class == AssetClass.CRYPTO:
        # For crypto, use UTC midnight of prior day
        utc_now = datetime.now(pytz.timezone("UTC"))
        target_date = utc_now.date() - timedelta(days=1)
    else:
        # For stocks, use today if available, otherwise last trading day
        if today_date in sorted_dates:
            target_date = today_date
        else:
            target_date = sorted_dates[-1] if sorted_dates else today_date

    print(f"ORB CALCULATOR: Target date: {target_date}")
    logger.info(f"ORB Calculator: Target date: {target_date}")
    
    if target_date == today_date:
        logger.info(f"ORB CALCULATOR: ✅ Using TODAY ({target_date}) as target date")
    else:
        days_behind = (today_date - target_date).days
        logger.warning(f"ORB CALCULATOR: ⚠️ Using {target_date} as target date ({days_behind} days behind today)")

    # Calculate average volume from past periods (excluding target date)
    target_volume_date = target_date if target_date in or_volumes else sorted_dates[-1]

    if len(sorted_dates) > avg_period:
        past_volumes = [or_volumes[d] for d in sorted_dates[-avg_period - 1 : -1]]
    else:
        past_volumes = [or_volumes[d] for d in sorted_dates[:-1]]

    print(f"ORB CALCULATOR: Using {len(past_volumes)} past days for average")
    print(
        f"ORB CALCULATOR: Past dates: {sorted_dates[-avg_period - 1 : -1] if len(sorted_dates) > avg_period else sorted_dates[:-1]}"
    )
    print(f"ORB CALCULATOR: Past volumes: {past_volumes}")
    logger.info(f"ORB Calculator: Using {len(past_volumes)} past days for average")
    logger.info(
        f"ORB Calculator: Past dates: {sorted_dates[-avg_period - 1 : -1] if len(sorted_dates) > avg_period else sorted_dates[:-1]}"
    )
    logger.info(f"ORB Calculator: Past volumes: {past_volumes}")

    avg_vol = sum(past_volumes) / len(past_volumes) if past_volumes else 0.0

    current_vol = or_volumes.get(target_volume_date, 0.0)

    print(f"ORB CALCULATOR: Current day ({target_volume_date}) volume: {current_vol}")
    print(f"ORB CALCULATOR: Average volume (last {len(past_volumes)} days): {avg_vol}")
    logger.info(f"ORB Calculator: Current day ({target_volume_date}) volume: {current_vol}")
    logger.info(f"ORB Calculator: Average volume (last {len(past_volumes)} days): {avg_vol}")

    rel_vol = (
        (current_vol / avg_vol * 100) if avg_vol > 0 else (float("inf") if current_vol > 0 else 0.0)
    )

    print(
        f"ORB CALCULATOR: Relative volume calculation: {current_vol} / {avg_vol} * 100 = {rel_vol}%"
    )
    logger.info(
        f"ORB Calculator: Relative volume calculation: {current_vol} / {avg_vol} * 100 = {rel_vol}%"
    )

    # Get direction for target date
    direction = or_directions.get(target_date, "doji")

    print(f"ORB CALCULATOR: Final result - rel_vol={rel_vol}%, direction={direction}")
    logger.info(f"ORB Calculator: Final result - rel_vol={rel_vol}%, direction={direction}")
    
    # Final summary with data freshness check
    logger.info("=" * 80)
    logger.info(f"ORB CALCULATOR: Final Summary for {symbol.ticker}")
    logger.info(f"ORB CALCULATOR: Target date used: {target_date}")
    logger.info(f"ORB CALCULATOR: Relative volume: {rel_vol}%")
    logger.info(f"ORB CALCULATOR: Direction: {direction}")
    
    if target_date == today_date:
        logger.info(f"ORB CALCULATOR: ✅ Using REAL-TIME data (today)")
    else:
        days_behind = (today_date - target_date).days
        logger.warning(f"ORB CALCULATOR: ⚠️ Using DELAYED data ({days_behind} days behind)")
    
    logger.info("=" * 80)

    # Calculate opening range high and low for current target date
    target_or_bar = None
    if target_date in daily_groups:
        day_bars_sorted = sorted(daily_groups[target_date], key=lambda b: b["timestamp"])
        
        # Determine opening range time based on asset class
        if symbol.asset_class == AssetClass.CRYPTO:
            # For crypto, use UTC midnight (00:00:00) as opening range, convert to ET
            utc_midnight = pytz.timezone("UTC").localize(
                datetime.combine(target_date, datetime.strptime("00:00", "%H:%M").time())
            )
            open_time = utc_midnight.astimezone(pytz.timezone("US/Eastern"))
        else:
            # For stocks, use 9:30 AM Eastern as opening range
            open_time = create_market_open_time(target_date, hour=9, minute=30)
        
        open_range_end = open_time + timedelta(minutes=or_minutes)
        
        # Find bars that start within the opening range
        or_candidates = [
            bar for bar in day_bars_sorted if open_time <= bar["timestamp"] < open_range_end
        ]
        
        if or_candidates:
            target_or_bar = or_candidates[0]
    
    or_high = target_or_bar["high"] if target_or_bar else None
    or_low = target_or_bar["low"] if target_or_bar else None

    return {
        "rel_vol": rel_vol,
        "direction": direction,
        "or_high": or_high,
        "or_low": or_low,
    }

File: nodes/orb-calculator.ts
// src/services/indicator-calculators/orb-calculator.ts
// Translated from: services/indicator_calculators/orb_calculator.py

import { AssetClass, type AssetSymbol, type OHLCVBar } from './types';

export interface OrbResult {
  rel_vol: number | null;
  direction: string | null;
  or_high: number | null;
  or_low: number | null;
  error?: string;
}

/**
 * Get date string in YYYY-MM-DD format for ET timezone.
 */
function getEtDateString(timestampMs: number): string {
  const date = new Date(timestampMs);
  const etFormatter = new Intl.DateTimeFormat('en-CA', {
    timeZone: 'America/New_York',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  });
  return etFormatter.format(date);
}

/**
 * Get current date in ET timezone.
 */
function getTodayEtDate(): string {
  const now = new Date();
  const etFormatter = new Intl.DateTimeFormat('en-CA', {
    timeZone: 'America/New_York',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  });
  return etFormatter.format(now);
}

/**
 * Check if a bar timestamp falls within the opening range.
 */
function isInOpeningRange(
  barTimestampMs: number,
  openTimeMs: number,
  openRangeEndMs: number
): boolean {
  return barTimestampMs >= openTimeMs && barTimestampMs < openRangeEndMs;
}

/**
 * Create market open time for a given date (9:30 AM ET for stocks).
 */
function createMarketOpenTimeMs(dateStr: string, hour: number = 9, minute: number = 30): number {
  // Parse the date string (YYYY-MM-DD)
  const [year, month, day] = dateStr.split('-').map(Number);

  // Create date at the specified time in local timezone, then adjust for ET
  // This is a simplified approach - in production you might want a proper timezone library
  const date = new Date(year, month - 1, day, hour, minute, 0, 0);

  // Get the offset for Eastern Time (accounting for DST)
  const etOffset = getEtOffset(date);

  // Adjust to get UTC time that represents the ET time
  return date.getTime() + (date.getTimezoneOffset() * 60000) + (etOffset * 3600000);
}

/**
 * Get Eastern Time offset from UTC in hours (accounting for DST).
 */
function getEtOffset(date: Date): number {
  // ET is UTC-5 (EST) or UTC-4 (EDT)
  const jan = new Date(date.getFullYear(), 0, 1);
  const jul = new Date(date.getFullYear(), 6, 1);

  // Determine if we're in DST based on the date
  const isDst = date.getTimezoneOffset() < Math.max(jan.getTimezoneOffset(), jul.getTimezoneOffset());

  return isDst ? -4 : -5;
}

interface BarData {
  timestamp: number; // milliseconds
  dateStr: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

/**
 * Calculate ORB (Opening Range Breakout) indicator including relative volume and direction.
 *
 * This implementation extracts opening range data from 5-minute bars and calculates:
 * 1. Relative Volume: Current opening range volume vs average of previous periods
 * 2. Direction: Bullish, bearish, or doji based on opening range price movement
 *
 * @param bars - List of 5-minute OHLCV bars (aggregated bars from Polygon)
 * @param symbol - Asset symbol (used to determine opening range time based on asset class)
 * @param orMinutes - Opening range period in minutes (default: 5)
 * @param avgPeriod - Period for calculating average volume (default: 14 days)
 * @param nowFunc - Optional function to use instead of Date.now for testing
 * @returns Dictionary with 'rel_vol', 'direction', 'or_high', and 'or_low'
 *
 * Reference Paper Implementation:
 *   The paper uses:
 *   - Relative Volume = (Volume in first n minutes) / (Avg of last 14 days) * 100
 */
export function calculateOrb(
  bars: readonly OHLCVBar[],
  symbol: AssetSymbol,
  orMinutes: number = 5,
  avgPeriod: number = 14,
  nowFunc: (() => number) | null = null
): OrbResult {
  if (!bars || bars.length === 0) {
    return {
      rel_vol: null,
      direction: null,
      or_high: null,
      or_low: null,
      error: 'No bars provided',
    };
  }

  // Convert bars to internal format with ET date strings
  const barData: BarData[] = bars.map((bar) => ({
    timestamp: bar.timestamp,
    dateStr: getEtDateString(bar.timestamp),
    open: bar.open,
    high: bar.high,
    low: bar.low,
    close: bar.close,
    volume: bar.volume ?? 0,
  }));

  // Get today's date in ET
  const todayDate = nowFunc ? getEtDateString(nowFunc()) : getTodayEtDate();

  // Group bars by date
  const dailyGroups: Map<string, BarData[]> = new Map();
  for (const bar of barData) {
    if (!dailyGroups.has(bar.dateStr)) {
      dailyGroups.set(bar.dateStr, []);
    }
    dailyGroups.get(bar.dateStr)!.push(bar);
  }

  // Calculate opening range volumes, directions, highs, and lows for each day
  const orVolumes: Map<string, number> = new Map();
  const orDirections: Map<string, string> = new Map();
  const orHighs: Map<string, number> = new Map();
  const orLows: Map<string, number> = new Map();

  for (const [dateKey, dayBars] of dailyGroups) {
    // Sort bars by timestamp to ensure chronological order
    const dayBarsSorted = [...dayBars].sort((a, b) => a.timestamp - b.timestamp);

    // Determine opening range time based on asset class
    let openTimeMs: number;
    if (symbol.assetClass === AssetClass.CRYPTO) {
      // For crypto, use UTC midnight (00:00:00)
      const [year, month, day] = dateKey.split('-').map(Number);
      openTimeMs = Date.UTC(year, month - 1, day, 0, 0, 0);
    } else {
      // For stocks, use 9:30 AM Eastern as opening range
      openTimeMs = createMarketOpenTimeMs(dateKey, 9, 30);
    }

    // Look for bars within the opening range (uses orMinutes parameter)
    const openRangeEndMs = openTimeMs + orMinutes * 60 * 1000;

    // Find bars that start within the opening range
    const orCandidates = dayBarsSorted.filter((bar) =>
      isInOpeningRange(bar.timestamp, openTimeMs, openRangeEndMs)
    );

    if (orCandidates.length === 0) {
      continue;
    }

    // Pick the earliest bar in the opening range for volume and direction
    const orBar = orCandidates[0];

    // Calculate opening range metrics from earliest bar
    const orVolume = orBar.volume;
    const orOpen = orBar.open;
    const orClose = orBar.close;

    // Calculate high/low across all opening range bars (handles orMinutes > 5)
    const orHigh = Math.max(...orCandidates.map((b) => b.high));
    const orLow = Math.min(...orCandidates.map((b) => b.low));

    // Determine direction
    let direction: string;
    if (orClose > orOpen) {
      direction = 'bullish';
    } else if (orClose < orOpen) {
      direction = 'bearish';
    } else {
      direction = 'doji';
    }

    orVolumes.set(dateKey, orVolume);
    orDirections.set(dateKey, direction);
    orHighs.set(dateKey, orHigh);
    orLows.set(dateKey, orLow);
  }

  // Calculate relative volume
  const sortedDates = [...orVolumes.keys()].sort();
  if (sortedDates.length < 1) {
    return {
      rel_vol: null,
      direction: null,
      or_high: null,
      or_low: null,
      error: 'Insufficient days',
    };
  }

  // Determine target date (today or last trading day)
  let targetDate: string;
  if (symbol.assetClass === AssetClass.CRYPTO) {
    // For crypto, use previous day
    const yesterday = new Date();
    yesterday.setDate(yesterday.getDate() - 1);
    targetDate = getEtDateString(yesterday.getTime());
  } else {
    // For stocks, use today if available, otherwise last trading day
    if (sortedDates.includes(todayDate)) {
      targetDate = todayDate;
    } else {
      targetDate = sortedDates[sortedDates.length - 1];
    }
  }

  // Calculate average volume from past periods (excluding target date)
  const targetVolumeDate = orVolumes.has(targetDate)
    ? targetDate
    : sortedDates[sortedDates.length - 1];

  let pastVolumes: number[];
  if (sortedDates.length > avgPeriod) {
    pastVolumes = sortedDates
      .slice(-avgPeriod - 1, -1)
      .map((d) => orVolumes.get(d)!)
      .filter((v) => v !== undefined);
  } else {
    pastVolumes = sortedDates
      .slice(0, -1)
      .map((d) => orVolumes.get(d)!)
      .filter((v) => v !== undefined);
  }

  const avgVol =
    pastVolumes.length > 0
      ? pastVolumes.reduce((a, b) => a + b, 0) / pastVolumes.length
      : 0.0;

  const currentVol = orVolumes.get(targetVolumeDate) ?? 0.0;

  let relVol: number;
  if (avgVol > 0) {
    relVol = (currentVol / avgVol) * 100;
  } else if (currentVol > 0) {
    relVol = Infinity;
  } else {
    relVol = 0.0;
  }

  // Get direction, high, and low for target date
  const direction = orDirections.get(targetDate) ?? 'doji';
  const orHigh = orHighs.get(targetDate) ?? null;
  const orLow = orLows.get(targetDate) ?? null;

  return {
    rel_vol: relVol,
    direction,
    or_high: orHigh,
    or_low: orLow,
  };
}
